src/Operation/ReplaceOne.php

```diff
@@ -76,6 +76,10 @@ public function __construct($databaseName, $collectionName, $filter, $replacemen
             throw new InvalidArgumentException('First key in $replacement argument is an update operator');
         }
 
+        if (\MongoDB\is_pipeline($replacement)) {
```

Comment:
I _think_ this makes sense, although I didn't find this discussed in the CRUD spec. In hindsight, I think https://github.com/mongodb/specifications/commit/7246066275c34a3609defd27527121a350cce3da#diff-5efd2e4ce61f4ccda0ce031099303996 for [SPEC-1267](https://jira.mongodb.org/browse/SPEC-1267) should have addressed this in the CRUD spec's [Update vs. Replace Validation](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#update-vs-replace-validation) section. It could also have added a failing test for `replaceOne` and replace via `BulkWrite`. Would you mind opening a spec ticket? Hopefully we can get to it during SPEC triage Monday afternoon (NYC time).

Although documents cannot contain keys prefixed with `$`, I was wondering if the following replacement document might be ambiguous (at least for our `is_pipeline()` function):

```php
[
    [ '$ref' => '<any string>' ],
    [ '$id' => '<any non-array value>' ],
]
```

This could be interpreted as a replacement document assigning a [DBRef](https://docs.mongodb.com/manual/reference/database-references/#dbrefs) to the document's top-level "0" field, or a pipeline with `$ref` and `$id` stages. I decided to test this in the `mongo` shell:

```
rs0:PRIMARY> db.foo.drop()
false

rs0:PRIMARY> db.foo.insert({"0":1})
WriteResult({ "nInserted" : 1 })

rs0:PRIMARY> db.foo.findOne()
{ "_id" : ObjectId("5d61d496637adc23cd5c4215"), "0" : 1 }

rs0:PRIMARY> db.foo.replaceOne({},[{"$ref":"foo","$id":"bar"}])
2019-08-24T20:26:51.491-0400 E  QUERY    [js] uncaught exception: Error: Cannot use pipeline-style updates in a replacement operation :
DBCollection.prototype.replaceOne@src/mongo/shell/crud_api.js:461:15
@(shell):1:1
```

That supports the validation you have here. Still, it's a client-side error and I was wondering what the server-side error might be if we actually passed this along to the `update` command. Does MongoDB 4.2 accept a BSON array as a valid replacement document in the [`update` command's `updates[].u` option](https://docs.mongodb.com/manual/reference/command/update/#update-command-u)? Recent server versions added stricter validation, but I believe in some older servers (perhaps in the 3.x range) a BSON array/document were accepted interchangeably. Let's find out:

```
rs0:PRIMARY> db.runCommand({update:"foo",updates:[{q:{},u:[{"$ref":"foo","$id":"bar"}]}]});
{
	"n" : 0,
	"nModified" : 0,
	"writeErrors" : [
		{
			"index" : 0,
			"code" : 40323,
			"errmsg" : "A pipeline stage specification object must contain exactly one field."
		}
	],
        ...
}
```

My DBRef example is a total edge case, but let's suppose an application actually did want to perform that replacement for some reason. Passing an actual BSON document (i.e. object cast on the PHP array) does the trick:

```
rs0:PRIMARY> db.runCommand({update:"foo",updates:[{q:{},u:{"0":{"$ref":"foo","$id":"bar"}}}]}); 
{
	"n" : 1,
	"nModified" : 1,
        ...
}
```

I also traced the code path for constructing the `update` command and we can see in [`_mongoc_write_command_update_append()`](https://github.com/mongodb/mongo-c-driver/blob/1.15.0/src/libmongoc/src/mongoc/mongoc-write-command.c#L121) that libmongoc always appends the `u` option as a document unless it's a pipeline. So even if a PHP array was used as a replacement document and managed to sneak past our update/replace validation, we could trust libmongoc to actually send a BSON document to the server.

Reply:
Thank you for that detailed analysis. I've created a spec ticket to ask for clarification in the spec. From the result of your tests, I deduce that our handling is correct in this case with users having to explicitly pass an object for a replacement that you suggested above.

I agree that it's an edge case, especially considering that DBRef objects normally contain both `$ref` and `$id` keys. With that in mind, we could add logic to `MongoDB\is_pipeline` to assert that each stage in the potential pipeline only contains a single element; this would at least remove the client-side assertion when replacing a document with a single DBRef object, as we would no longer accept that as a valid aggregation pipeline. What do you think?